api/broker/broker_controller.go

```diff
@@ -85,94 +97,144 @@ func (ctrl *Controller) addBroker(response http.ResponseWriter, request *http.Re
 	err = common.HandleNotFoundError(err, "broker", brokerID)
 	if err != nil {
 		return err
 	}
-	broker.Credentials = nil
 
+	broker.Credentials = nil
+	broker.Catalog = nil
 	return rest.SendJSON(response, http.StatusOK, broker)
 }
 
-func (ctrl *Controller) getAllBrokers(response http.ResponseWriter, request *http.Request) error {
+func (c *Controller) getAllBrokers(response http.ResponseWriter, request *http.Request) error {
 	logrus.Debug("Getting all brokers")
-
-	brokers, err := ctrl.BrokerStorage.GetAll()
+	brokers, err := c.BrokerStorage.GetAll()
 	if err != nil {
 		return err
 	}
+	withCatalog := request.FormValue(catalogParam)
+	if strings.ToLower(withCatalog) != "true" {
+		for i := 0; i < len(brokers); i++ {
+			brokers[i].Catalog = nil
+		}
+	}
 
-	type brokerResponse struct {
+	return rest.SendJSON(response, http.StatusOK, struct {
 		Brokers []types.Broker `json:"brokers"`
-	}
-	return rest.SendJSON(response, http.StatusOK, brokerResponse{brokers})
+	}{
+		Brokers: brokers,
+	})
 }
 
-func (ctrl *Controller) deleteBroker(response http.ResponseWriter, request *http.Request) error {
+func (c *Controller) deleteBroker(response http.ResponseWriter, request *http.Request) error {
 	brokerID := getBrokerID(request)
 	logrus.Debugf("Deleting broker with id %s", brokerID)
 
-	err := ctrl.BrokerStorage.Delete(brokerID)
+	err := c.BrokerStorage.Delete(brokerID)
 	err = common.HandleNotFoundError(err, "broker", brokerID)
 	if err != nil {
 		return err
 	}
 	return rest.SendJSON(response, http.StatusOK, map[string]int{})
 }
 
-func (ctrl *Controller) updateBroker(response http.ResponseWriter, request *http.Request) error {
+func (c *Controller) patchBroker(response http.ResponseWriter, request *http.Request) error {
 	brokerID := getBrokerID(request)
-	logrus.Debugf("Updating broker with id %s", brokerID)
+	logrus.Debugf("Updating updateBroker with id %s", brokerID)
 
-	broker := &types.Broker{}
-	if err := rest.ReadJSONBody(request, broker); err != nil {
-		logrus.Error("Invalid request body")
+	updateBroker := &types.Broker{}
+	if err := rest.ReadJSONBody(request, updateBroker); err != nil {
 		return err
 	}
 
-	broker.ID = brokerID
-	broker.UpdatedAt = time.Now().UTC()
+	updateBroker.ID = brokerID
+	updateBroker.UpdatedAt = time.Now().UTC()
 
-	brokerStorage := ctrl.BrokerStorage
-	if broker.Credentials != nil {
-		err := validateBrokerCredentials(broker.Credentials)
+	if updateBroker.Credentials != nil {
+		err := validateBrokerCredentials(updateBroker.Credentials)
 		if err != nil {
 			return types.NewErrorResponse(err, http.StatusBadRequest, "BadRequest")
 		}
 	}
-	err := brokerStorage.Update(broker)
+
+	broker, err := c.BrokerStorage.Get(updateBroker.ID)
+	err = common.HandleNotFoundError(err, "broker", brokerID)
+	if err != nil {
+		return err
+	}
+
+	updateData, err := json.Marshal(updateBroker)
+	if err != nil {
+		return err
+	}
+	if err := json.Unmarshal(updateData, broker); err != nil {
```

Comment:
There seems to be some problem with the Unmarshaling and timestamps (I think?) The "created_at" which is returned as a response is always "0001-01-01T00:00:00Z". The value in the DB is as expected. Maybe the whole marshal/unmarshal could be removed.

Reply:
I am not sure that it can be removed - we need some way to get the whole updated broker. 
There are two reasons we need the updated broker 
1. As per specification we want to return it in the response (this is easily changeable and we may return empty response too). (Basically we need to merge the updated broker with the broker that we already obtained from db)
2. let's say the request contains only a new updated brokerURL - getBrokerCatalog needs more info than just URL. So we need a smart way to merge the broker from db with the request object. 

I will check what's up with the created field.

Reply:
Why don't we first update the broker in the database, then do a get broker from the database, they will be already merged. Then with this information we can do catalog fetch and return the updated broker to the client

Reply:
If the catalog fetch fails we would need to do delete broker from the database. It would be better if we can avoid it (i.e. save in db only if fetching of catalog was successful).